source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,176 @@
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // The values in the |message_.result_id_map| must be all fresh and all
+  // distinct.
+  std::set<uint32_t> ids_used_by_this_transformation;
+  for (auto& pair : message_.result_id_map()) {
+    if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+            pair.second, ir_context, &ids_used_by_this_transformation)) {
+      return false;
+    }
+  }
+
+  // |function_call_instruction| must be defined and must be an OpFunctionCall
+  // instruction.
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  if (function_call_instruction == nullptr ||
+      function_call_instruction->opcode() != SpvOpFunctionCall) {
+    return false;
+  }
+
+  // The called function must not have an early return.
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+  if (called_function->HasEarlyReturn()) {
+    return false;
+  }
+
+  // |message_.result_id_map| must have an entry for every result id in the
+  // called function.
+  for (auto& block : *called_function) {
+    // Since the entry block label will not be inlined, only the remaining
+    // labels must have a corresponding value in the map.
+    if (&block != &*called_function->entry() &&
+        !message_.result_id_map().count(block.GetLabel()->result_id())) {
+      return false;
+    }
+    for (auto& instruction : block) {
+      if (instruction.HasResultId() &&
+          instruction.result_id() != called_function->GetReturnValueId() &&
+          !message_.result_id_map().count(instruction.result_id())) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+void TransformationInlineFunction::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  auto caller_function =
+      ir_context->get_instr_block(function_call_instruction)->GetParent();
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+
+  for (auto& block : *called_function) {
+    // The called function entry block label will not be inlined.
+    if (&block != &*called_function->entry()) {
+      auto cloned_label_instruction = block.GetLabelInst()->Clone(ir_context);
+      uint32_t fresh_id =
+          message_.result_id_map().at(cloned_label_instruction->result_id());
+      cloned_label_instruction->SetResultId(fresh_id);
+      function_call_instruction->InsertBefore(
+          std::unique_ptr<opt::Instruction>(cloned_label_instruction));
+      fuzzerutil::UpdateModuleIdBound(ir_context, fresh_id);
+    }
+
+    for (auto& instruction : block) {
+      // The function return instruction will not be inlined.
+      if (&instruction == &*called_function->tail()->tail()) {
```

Comment:
This isn't the right way to identify the function return instruction because it might not be the final instruction in the function.  Remember that it can be anywhere in the function so long as dominance rules are respected.

I assume you're going to change the single "OpReturn" or "OpReturnValue" in the function into an "OpBranch" to the basic block that follows the block containing the function call?

Reply:
I guess the return instruction must be changed to the last instruction in the block containing the function call instruction.
I'll leave this changes to the next commit so that we can talk about this and the last comment.

Reply:
@andreperezmaselco:

- OpReturn should turn into OpBranch (to the block following the original function call)

- OpReturnValue %v should turn into %w = OpCopyObject %v; OpBranch (to the block following the original function call)
where %w is the receiving variable for the original OpFunctionCall instruction.

Reply:
Done.